lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -63,6 +59,94 @@ struct ConnectInfo {
 using InnerRefMap =
     DenseMap<hw::InnerRefAttr, std::pair<hw::InnerSymAttr, StringAttr>>;
 
+//===----------------------------------------------------------------------===//
+// Naming Helpers
+//===----------------------------------------------------------------------===//
+
+/// For a layer `@A::@B::@C` in module Module,
+/// the generated module is called `Module_A_B_C`.
+static SmallString<32> moduleNameForLayer(StringRef moduleName,
+                                          SymbolRefAttr layerName) {
+  SmallString<32> result;
+  bool dirty = false;
+  if (!moduleName.empty()) {
+    result.append(moduleName);
+    dirty = true;
+  }
+  auto root = layerName.getRootReference();
+  if (!root.empty()) {
+    if (dirty)
+      result.append("_");
+    result.append(root);
+    dirty = true;
+  }
+  for (auto nested : layerName.getNestedReferences()) {
+    auto part = nested.getValue();
+    if (!part.empty()) {
+      if (dirty)
+        result.append("_");
+      result.append(part);
+      dirty = true;
+    }
+  }
+  return result;
+}
+
+/// For a layerblock `@A::@B::@C`,
+/// the generated instance is called `a_b_c`.
+static SmallString<32> instanceNameForLayer(SymbolRefAttr layerName) {
+  SmallString<32> result;
+  bool dirty = false;
+  auto root = layerName.getRootReference();
+  if (!root.empty()) {
+    result.append(root);
+    result[0] = llvm::toLower(result[0]);
+    dirty = true;
+  }
+  for (auto nested : layerName.getNestedReferences()) {
+    auto part = nested.getValue();
+    if (!part.empty()) {
+      if (dirty)
+        result.append("_");
+      auto i = result.size();
+      result.append(part);
+      result[i] = llvm::toLower(result[i]);
+      dirty = true;
+    }
+  }
+  return result;
+}
+
+/// For all layerblocks `@A::@B::@C` in a circuit called Circuit,
+/// the output filename is `layers_Circuit_A_B_C.sv`.
+static SmallString<32> fileNameForLayer(StringRef circuitName,
+                                        SymbolRefAttr layerName) {
+  SmallString<32> result;
+  result.append("layers");
+  if (!circuitName.empty()) {
+    result.append("_");
+    result.append(circuitName);
+  }
+  auto root = layerName.getRootReference();
+  if (!root.empty()) {
+    result.append("_");
+    result.append(root);
+  }
+  for (auto nested : layerName.getNestedReferences()) {
+    auto part = nested.getValue();
+    if (!part.empty()) {
+      result.append("_");
+      result.append(part);
+    }
+  }
+  result.append(".sv");
+  return result;
+}
```

Comment:
It seems like there is a missing, common utility that does the logic of converging the `SymbolRefAttr` to a string. Would this help clean things up?

``` c++
static void layerToString(SymbolRefAttr layerName, SmallString<32> &str);
```

It _might_ be cleaner to do this by defining the streaming operator `<<` on a symbol and using streaming to build up the string.

Reply:
Added some more helpers